cpp/src/arrow/util/config.h.cmake

```diff
@@ -57,6 +57,7 @@
 #cmakedefine ARROW_GCS
 #cmakedefine ARROW_HDFS
 #cmakedefine ARROW_S3
+#cmakedefine ARROW_TEST_MEMCHECK
```

Comment:
We already have `ARROW_VALGRIND`, I don't understand why we would need to introduce this new constant.

Reply:
Oh, sorry. I missed it. Can we move `ARROW_VALGRIND` to `config.h.cmake` from `add_definitions(-DARROW_VALGRIND)`?

Reply:
This is not an information that is useful to third-party code, so I don't understand why that would be useful.

Reply:
This will not be useful to third-party code but it helps us to know what macros are defined in a build.
If we use `add_definitions(-DARROW_VALGRIND)`, we need to use `ninja -v` and find a `-DARROW_VALGRIND` from a long command line.

Reply:
#41176/#41177